backends/edk2-compat/edk2-svc-generate.c

```diff
@@ -232,200 +206,148 @@ int performGenerateCommand(int argc,char* argv[])
 		free(args.signCerts);
 	if (args.time) 
 		free(args.time);
-	if (rc) 
-		printf("RESULT: FAILURE\n");
-	else 
-		printf("RESULT: SUCCESS\n");
+	if (!args.helpFlag) 
+		printf("RESULT: %s\n", rc ? "FAILURE" : "SUCCESS");
 	
 	return rc;
 }
 
 /**
- *@param argv , array of command line arguments
- *@param argc, length of argv
- *@param args, struct that will be filled with data from argv
+ *@param key , every option that is parsed has a value to identify it
+ *@param arg, if key is an option than arg will hold its value ex: -<key> <arg>
+ *@param state,  argp_state struct that contains useful information about the current parsing state 
  *@return success or errno
  */
-static int parseArgs( int argc, char *argv[], struct Arguments *args) {
+static int parse_opt(int key, char *arg, struct argp_state *state) 
+{
+	struct Arguments *args = state->input;
 	int rc = SUCCESS;
-	for (int i = 0; i < argc; i++) {
-		if (argv[i][0] == '-') {
-			if (!strcmp(argv[i], "--usage")) {
-				usage();
+	//this checks to see if help/usage is requested
+	//argp can either exit() or raise no errors, we want to go to cleanup and then exit so we need a special flag
+	//this becomes extra sticky since --usage/--help never actually get passed to this function
+	if (args->helpFlag == 0) {
+		if (state->next == 0 && state->next + 1 < state->argc) {
+			if (strncmp("--u", state->argv[state->next + 1], strlen("--u")) == 0 
+				|| strncmp("--h", state->argv[state->next + 1], strlen("--h")) == 0
+				|| strncmp("-?", state->argv[state->next + 1], strlen("-?")) == 0)
 				args->helpFlag = 1;
-				goto out;
-			}
-			else if (!strcmp(argv[i], "--help")) {
-				help();
+		}
+		else if (state->next < state->argc)
+			if (strncmp("--u", state->argv[state->next], strlen("--u")) == 0 
+				|| strncmp("--h", state->argv[state->next], strlen("--h")) == 0
+				|| strncmp("-?", state->argv[state->next], strlen("-?")) == 0)
 				args->helpFlag = 1;
```

Comment:
man, again, what a bummer.

I think we can make it a bit simpler by defining `--help` and `--usage` ourselves and have them invoke the functions argp provides for printing help and usage. (I just got this from skimming argp.h.)

There's a trick needed for having `usage` with just a long option which I got from https://stackoverflow.com/questions/47727755/gnu-argp-how-to-parse-option-with-only-long-name.

This allows us to clean things up quite a bit, and AFAICT the only downside is that `--usage` gets sorted before `--verbose` in the help messages. I think we can live with that.

See the attached diff.

[generate.diff.txt](https://github.com/open-power/secvarctl/files/6243767/generate.diff.txt)

If this approach seems good to you (and it's possible I missed some other subtlety) then it should apply to the other subcommands as well.

Reply:
(perhaps it's conventional for --help to go to stdout, you could check the argp source to be sure.)

Reply:
Added a new commit for these edits. I also figured out how to do the ordering of the arguments in the help message so that `usage` and `help` are together. Thanks a bunch for figuring this out. It looks a lot cleaner